src/contracts/middleware/IndexRegistry.sol

```diff
@@ -0,0 +1,188 @@
+
+        //blocknumber must be before the "index'th" entry's toBlockNumber
+        require(blockNumber <= operatorIndexToCheck.toBlockNumber, "IndexRegistry.getOperatorIndexForQuorumAtBlockNumberByIndex: provided index is too far in the future for provided block number");
+       
+        //if there is an index update before the "index'th" update, the blocknumber must be after than the previous entry's toBlockNumber
+        if(index != 0){
+            OperatorIndex memory previousOperatorIndex = operatorIdToIndexHistory[operatorId][quorumNumber][index - 1];
+            require(blockNumber > previousOperatorIndex.toBlockNumber, "IndexRegistry.getOperatorIndexForQuorumAtBlockNumberByIndex: provided index is too far in the past for provided block number");
+        }
+        return operatorIndexToCheck.index;
+    }
+
+    /**
+     * @notice Looks up the number of total operators for `quorumNumber` at the specified `blockNumber`.
+     * @param quorumNumber is the quorum number for which the total number of operators is desired
+     * @param blockNumber is the block number at which the total number of operators is desired
+     * @param index is the index of the entry in the dynamic array `totalOperatorsHistory[quorumNumber]` to read data from
+     */
+    function getTotalOperatorsForQuorumAtBlockNumberByIndex(uint8 quorumNumber, uint32 blockNumber, uint32 index) external view returns (uint32){
+        OperatorIndex memory operatorIndexToCheck = totalOperatorsHistory[quorumNumber][index];
+
+        //blocknumber must be before the "index'th" entry's toBlockNumber
+        require(blockNumber <= operatorIndexToCheck.toBlockNumber, "IndexRegistry.getTotalOperatorsForQuorumAtBlockNumberByIndex: provided index is too far in the future for provided block number");
+        
+        //if there is an index update before the "index'th" update, the blocknumber must be after than the previous entry's toBlockNumber
+        if (index != 0){
+            OperatorIndex memory previousOperatorIndex = totalOperatorsHistory[quorumNumber][index - 1];
+            require(blockNumber > previousOperatorIndex.toBlockNumber, "IndexRegistry.getTotalOperatorsForQuorumAtBlockNumberByIndex: provided index is too far in the past for provided block number");
+        }
+        return operatorIndexToCheck.index;
+    }
+
+    function totalOperatorsForQuorum(uint8 quorumNumber) external view returns (uint32){
+        return uint32(quorumToOperatorList[quorumNumber].length);
+    }
+
+
+    /// @notice Returns the current number of operators of this service.
+    function totalOperators() external view returns (uint32){
+        return uint32(globalOperatorList.length);
+    }
+
+
+    function _updateTotalOperatorHistory(uint8 quorumNumber) internal {
+
+        //if there is a prior entry, update its "toBlockNumber"
+        if (totalOperatorsHistory[quorumNumber].length > 0) {
```

Comment:
we prematurely load the length and store in memory in other places, we should do it here too

Reply:
done